modules/core/src/main/java/org/apache/ignite/internal/processors/cache/WalStateManager.java

```diff
@@ -1013,12 +1015,14 @@ private WalStateResult awaitCheckpoint(CheckpointProgress cpFut, WalStatePropose
      * Checks WAL disabled for cache group.
      *
      * @param grpId Group id.
+     * @param pageId Page id.
      * @return {@code True} if WAL disable for group. {@code False} If not.
      */
-    public boolean isDisabled(int grpId) {
+    public boolean isDisabled(int grpId, long pageId) {
         CacheGroupContext ctx = cctx.cache().cacheGroup(grpId);
 
-        return ctx != null && !ctx.walEnabled();
+        return ctx != null && (!ctx.walEnabled()
+            || (!ctx.indexWalEnabled() && PageIdUtils.partId(pageId) == INDEX_PARTITION));
```

Comment:
Relocation of `PageIdUtils.partId(pageId) == INDEX_PARTITION))` to the `ctx.indexWalEnabled()` should simplify the code.